File: packages/dart/test/obj_export_test.dart
import 'dart:typed_data';
import 'package:openskp/openskp.dart';
import 'package:test/test.dart';

void main() {
  group('Wavefront OBJ Exporter', () {
    test('serializes Scene to OBJ text format', () {
      final scene = Scene(
        sceneHierarchy: InstanceNode(
          name: 'Root',
          definitionName: 'RootDef',
          layer: 'Layer0',
          positionMm: (0.0, 0.0, 0.0),
          properties: {},
          children: [],
        ),
        meshIndex: {},
        glbPrimitives: [
          GlbPrimitive(
            geomName: 'Cube',
            materialIndex: 0,
            positions: Float32List.fromList([0, 0, 0, 1, 0, 0, 0, 1, 0]),
            normals: Float32List.fromList([0, 0, 1, 0, 0, 1, 0, 0, 1]),
            uvs: Float32List.fromList([0, 0, 1, 0, 0, 1]),
            indices: Uint32List.fromList([0, 1, 2]),
          ),
        ],
        gltfMaterials: [],
      );

      final objText = toObj(scene);
      expect(objText, contains('# OpenSKP OBJ Export'));
      expect(objText, contains('o Cube'));
      expect(objText, contains('v 0.000000 0.000000 0.000000'));
      expect(objText, contains('f 1 2 3'));
    });
  });
}

File: packages/cpp/tests/obj_export_test.cpp
#include "openskp/obj_export.hpp"

#include <gtest/gtest.h>

namespace openskp {
namespace {

TEST(ObjExport, SerializesSceneToObjText) {
  Scene scene;
  GlbPrimitive prim;
  prim.geom_name = "Cube";
  prim.material_index = 0;
  prim.positions = {0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0};
  prim.normals = {0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0};
  prim.uvs = {0.0, 0.0, 1.0, 0.0, 0.0, 1.0};
  prim.indices = {0, 1, 2};
  scene.glb_primitives.push_back(prim);

  std::string obj_text = to_obj(scene);
  EXPECT_NE(obj_text.find("# OpenSKP OBJ Export"), std::string::npos);
  EXPECT_NE(obj_text.find("o Cube"), std::string::npos);
  EXPECT_NE(obj_text.find("v 0.000000 0.000000 0.000000"), std::string::npos);
  EXPECT_NE(obj_text.find("f 1 2 3"), std::string::npos);
}

}  // namespace
}  // namespace openskp
